File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/CategoryFragment.java
package com.remotearthsolutions.expensetracker.fragments;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.activities.ApplicationObject;
import com.remotearthsolutions.expensetracker.adapters.CategoryListViewAdapter;
import com.remotearthsolutions.expensetracker.contracts.CategoryFragmentContract;
import com.remotearthsolutions.expensetracker.databaseutils.DatabaseClient;
import com.remotearthsolutions.expensetracker.databaseutils.daos.CategoryDao;
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel;
import com.remotearthsolutions.expensetracker.viewmodels.CategoryViewModel;

import java.util.List;

public class CategoryFragment extends BaseFragment implements CategoryFragmentContract.View, OptionBottomSheetFragment.Callback {

    private RecyclerView recyclerView;
    private FloatingActionButton floatingActionButton;
    private CategoryListViewAdapter adapter;
    private CategoryViewModel viewModel;
    private CategoryModel selectedCategory;
    private int limitOfCategory;


    public CategoryFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_category, container, false);

        floatingActionButton = view.findViewById(R.id.addcategory);
        recyclerView = view.findViewById(R.id.cat_recycler);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));

        CategoryDao categoryDao = DatabaseClient.getInstance(getContext()).getAppDatabase().categoryDao();
        viewModel = new CategoryViewModel(this, categoryDao);
        viewModel.showCategories();

        viewModel.getNumberOfItem().observe(getViewLifecycleOwner(), (Integer count) -> limitOfCategory = count);

        floatingActionButton.setOnClickListener(v -> {
            if (limitOfCategory < 20 ||
                    ((ApplicationObject) getActivity().getApplication()).isPremium()) {
                selectedCategory = null;
                onClickEditBtn();
            } else {
                showAlert(getString(R.string.attention), getString(R.string.you_need_to_be_premium_user_to_add_more_categories), getString(R.string.ok), null, null);
            }
        });

        return view;
    }

    @Override
    public void showCategories(List<CategoryModel> categories) {

        adapter = new CategoryListViewAdapter(categories);
        adapter.setOnItemClickListener(categoryModel -> {

            selectedCategory = categoryModel;
            OptionBottomSheetFragment optionBottomSheetFragment = new OptionBottomSheetFragment();
            optionBottomSheetFragment.setCallback(CategoryFragment.this, OptionBottomSheetFragment.OptionsFor.CATEGORY);
            optionBottomSheetFragment.show(getChildFragmentManager(), OptionBottomSheetFragment.class.getName());

        });
        recyclerView.setAdapter(adapter);

    }

    @Override
    public void onClickAddAmountBtn() {
        //THis is not need for category. need to refactor this somehow so this method will not be needed to implement here.
    }

    @Override
    public void onClickEditBtn() {
        FragmentManager fm = getChildFragmentManager();
        final AddCategoryDialogFragment categoryDialogFragment = AddCategoryDialogFragment.newInstance(getString(R.string.update_category));
        categoryDialogFragment.setCategory(selectedCategory);
        categoryDialogFragment.setCallback(categoryModel1 -> {
            //viewModel.showCategories();
            categoryDialogFragment.dismiss();

        });
        categoryDialogFragment.show(fm, AddCategoryDialogFragment.class.getName());
    }

    @Override
    public void onClickDeleteBtn() {

        if (selectedCategory.getNotremovable() == 1) {
            showToast(getString(R.string.you_cannot_delete_this_category));
            return;
        }
        showAlert(getString(R.string.warning), getString(R.string.deleting_this_category_will_remove_expenses_related_to_this_also_are_you_sure_you_want_to_delete), getString(R.string.yes), getString(R.string.not_now), new Callback() {
            @Override
            public void onOkBtnPressed() {
                viewModel.deleteCategory(selectedCategory);
                Toast.makeText(getActivity(), getString(R.string.category_deleted_successfully), Toast.LENGTH_LONG).show();
            }

            @Override
            public void onCancelBtnPressed() {

            }
        });
    }

    @Override
    public Context getContext() {
        return getActivity();
    }
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/CategoryFragment.kt
package com.remotearthsolutions.expensetracker.fragments

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.activities.ApplicationObject
import com.remotearthsolutions.expensetracker.activities.main.MainActivity
import com.remotearthsolutions.expensetracker.adapters.CategoryListViewAdapter
import com.remotearthsolutions.expensetracker.contracts.BaseView
import com.remotearthsolutions.expensetracker.contracts.CategoryFragmentContract
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel
import com.remotearthsolutions.expensetracker.databinding.FragmentCategoryBinding
import com.remotearthsolutions.expensetracker.fragments.OptionBottomSheetFragment.OptionsFor
import com.remotearthsolutions.expensetracker.utils.AnalyticsManager
import com.remotearthsolutions.expensetracker.viewmodels.CategoryViewModel
import org.koin.androidx.viewmodel.ext.android.viewModel
import org.koin.core.parameter.parametersOf

class CategoryFragment : BaseFragment(),
    CategoryFragmentContract.View,
    OptionBottomSheetFragment.Callback {
    private lateinit var binding: FragmentCategoryBinding
    private lateinit var adapter: CategoryListViewAdapter
    private val viewModel: CategoryViewModel by viewModel { parametersOf(this) }
    private var selectedCategory: CategoryModel? = null
    private var limitOfCategory = 0
    private var mContext: Context? = null
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentCategoryBinding.inflate(layoutInflater, container, false)
        binding.catRecycler.setHasFixedSize(true)
        binding.catRecycler.layoutManager = LinearLayoutManager(mContext)

        viewModel.showCategories()
        viewModel.numberOfItem.observe(
            viewLifecycleOwner
        ) { count: Int -> limitOfCategory = count }

        binding.addcategory.setOnClickListener {
            if (limitOfCategory < 20 ||
                ((mContext as Activity?)!!.application as ApplicationObject).isPremium
            ) {
                selectedCategory = null
                onClickEditBtn()
            } else {
                showAlert(
                    getResourceString(R.string.attention),
                    getResourceString(R.string.you_need_to_be_premium_user_to_add_more_categories),
                    getResourceString(R.string.ok),
                    null,
                    null, null
                )
            }
        }
        registerBackButton()
        return binding.root
    }

    override fun showCategories(categories: List<CategoryModel>?) {
        adapter = CategoryListViewAdapter(categories)
        adapter.setOnItemClickListener(object : CategoryListViewAdapter.OnItemClickListener {
            override fun onItemClick(categoryModel: CategoryModel?) {
                selectedCategory = categoryModel
                val optionBottomSheetFragment =
                    OptionBottomSheetFragment()
                optionBottomSheetFragment.setCallback(this@CategoryFragment, OptionsFor.CATEGORY)
                optionBottomSheetFragment.show(
                    childFragmentManager,
                    OptionBottomSheetFragment::class.java.name
                )
            }
        })
        binding.catRecycler.adapter = adapter
    }

    override fun onClickEditBtn() {
        val fm = childFragmentManager
        val categoryDialogFragment: AddCategoryDialogFragment =
            AddCategoryDialogFragment.newInstance(getResourceString(R.string.update_category))
        categoryDialogFragment.setCategory(selectedCategory)
        categoryDialogFragment.setCallback(object : AddCategoryDialogFragment.Callback {
            override fun onCategoryAdded(categoryModel: CategoryModel?) {
                viewModel.showCategories()
                categoryDialogFragment.dismiss()
                (activity as MainActivity).onUpdateCategory()
            }
        })
        categoryDialogFragment.show(fm, AddCategoryDialogFragment::class.java.name)
    }

    override fun onClickDeleteBtn() {
        if (selectedCategory!!.notremovable == 1) {
            showToast(getResourceString(R.string.you_cannot_delete_this_category))
            return
        }
        showAlert(
            getResourceString(R.string.warning),
            getResourceString(R.string.deleting_this_category_will_remove_expenses_related_to_this_also_are_you_sure_you_want_to_delete),
            getResourceString(R.string.yes),
            getResourceString(R.string.not_now),
            null,
            object : BaseView.Callback {
                override fun onOkBtnPressed() {
                    viewModel.deleteCategory(selectedCategory!!)
                    Toast.makeText(
                        activity,
                        getResourceString(R.string.category_deleted_successfully),
                        Toast.LENGTH_LONG
                    ).show()
                    (activity as MainActivity).onUpdateCategory()
                    with(AnalyticsManager) {
                        logEvent(EXPENSE_CATEGORY_DELETED)
                    }
                }

                override fun onCancelBtnPressed() {}
            })
    }
}
